src/include/duckdb/common/types/row/tuple_data_collection.hpp

```diff
@@ -0,0 +1,237 @@
+	                      TupleDataPinProperties properties = TupleDataPinProperties::UNPIN_AFTER_DONE);
+	//! Initializes an Append state - useful for optimizing many appends made to the same tuple data collection
+	void InitializeAppend(TupleDataAppendState &append_state, vector<column_t> column_ids,
+	                      TupleDataPinProperties properties = TupleDataPinProperties::UNPIN_AFTER_DONE);
+	//! Initializes the Pin state of an Append state
+	//! - Useful for optimizing many appends made to the same tuple data collection
+	void InitializeAppend(TupleDataPinState &pin_state,
+	                      TupleDataPinProperties = TupleDataPinProperties::UNPIN_AFTER_DONE);
+	//! Initializes the Chunk state of an Append state
+	//! - Useful for optimizing many appends made to the same tuple data collection
+	void InitializeAppend(TupleDataChunkState &chunk_state, vector<column_t> column_ids);
+	//! Append a DataChunk directly to this TupleDataCollection - calls InitializeAppend and Append internally
+	void Append(DataChunk &new_chunk, const SelectionVector &append_sel = *FlatVector::IncrementalSelectionVector(),
+	            idx_t append_count = DConstants::INVALID_INDEX);
+	//! Append a DataChunk directly to this TupleDataCollection - calls InitializeAppend and Append internally
+	void Append(DataChunk &new_chunk, vector<column_t> column_ids,
+	            const SelectionVector &append_sel = *FlatVector::IncrementalSelectionVector(),
+	            const idx_t append_count = DConstants::INVALID_INDEX);
+	//! Append a DataChunk to this TupleDataCollection using the specified Append state
+	void Append(TupleDataAppendState &append_state, DataChunk &new_chunk,
+	            const SelectionVector &append_sel = *FlatVector::IncrementalSelectionVector(),
+	            const idx_t append_count = DConstants::INVALID_INDEX);
+	//! Append a DataChunk to this TupleDataCollection using the specified pin and Chunk states
+	void Append(TupleDataPinState &pin_state, TupleDataChunkState &chunk_state, DataChunk &new_chunk,
+	            const SelectionVector &append_sel = *FlatVector::IncrementalSelectionVector(),
+	            const idx_t append_count = DConstants::INVALID_INDEX);
+	//! Append a DataChunk to this TupleDataCollection using the specified pin and Chunk states
+	//! - ToUnifiedFormat has already been called
+	void AppendUnified(TupleDataPinState &pin_state, TupleDataChunkState &chunk_state, DataChunk &new_chunk,
+	                   const SelectionVector &append_sel = *FlatVector::IncrementalSelectionVector(),
+	                   const idx_t append_count = DConstants::INVALID_INDEX);
+
+	//! Creates a UnifiedVectorFormat in the given Chunk state for the given DataChunk
+	static void ToUnifiedFormat(TupleDataChunkState &chunk_state, DataChunk &new_chunk);
+	//! Gets the UnifiedVectorFormat from the Chunk state as an array
+	static unique_ptr<UnifiedVectorFormat[]> GetVectorData(const TupleDataChunkState &chunk_state);
```

Comment:
This forces the `UnifiedVectorFormat[]` to be allocated for each call - likely not very significant but we could take it as a parameter instead and cache it (which is what we do in the `AggregateHashtable` now).